Adding wal filter to inspect and filter wal records on recovery

PraveenSinghRao opened the pull request:
Adding the ability to supply a filter object to be invoked while processing write-ahead-logs during recovery. 

The filter provides a way to inspect log records, ignoring a particular record or skipping replay.

dhruba commented:
This patch is cool!

1. Would it be better have  another option to SkipProcessing all records till the end of the log?
2. Also, can you pl put this ifndef ROCKSDB_LITE so that we do not include this feature for android/ios?
3. Would it be possible to submit an "arc diff" so that we can provide review comments better?

dhruba commented:
4. I think u already have the option called kStopReplay to achieve what I wanted. Maybe what I want is an option to indicate that log-processing has encountered an error and the DB should not be opened.

5. Would it be good to send in a std::vector<WriteBatch> to the wal-filter so that the filter can process multiple log-records in one invocation? Such a feature might enable some kind of optimization in a custom-wal-filter?

PraveenSinghRao commented:
Done moving it under #ifndef ROCKSDB_LITE.

1. Maybe I can add an enum that returns kCorruptedRecord and it can be handled similar to the way corruption is handled without filter (based on an option).

2. This might need reworking of the code as single record nicely fits into the existing code. Is this something that can be added later?

I will try to see if I can get arcanist work on Windows - I believe others had some problem with it earlier.

siying commented on db/db_impl.cc:
```diff
@@ -1153,6 +1154,28 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& log_numbers,
 
+#ifndef ROCKSDB_LITE
+      if (db_options_.wal_filter != nullptr) {
+        WALFilter::WALProcessingOption walProcessingOption =
```
There are both of naming like "WALFilter" and "WalFilter" in the code base. We should consolidate it some time. "Wal" seems to be more to me so I would vote "WalFilter".

PraveenSinghRao replied:
For acronyms, all caps also seems fine to me, although I can change to whichever you prefer.

siying commented on include/rocksdb/options.h:
```diff
@@ -1130,6 +1131,15 @@ struct DBOptions {
+  // records, ignoring a particular record or skipping replay.
+  // The filter is invoked at startup and is invoked from a single-thread
+  // currently.
+  const WALFilter * wal_filter;
```
Nit: WALFilter* wal_filter

siying commented on include/rocksdb/wal_filter.h:
```diff
@@ -0,0 +1,67 @@
+};
+
+// Default implementation of WALFilter that does not alter WAL processing
+class DefaultWALFilter : WALFilter {
```
Do we need this default implementation at all?

PraveenSinghRao replied:
I saw something like this for compaction_filter and followed the pattern. Should this be removed?

siying replied:
That is legacy. We should remove that one too:) Let me just do it.

siying commented on include/rocksdb/wal_filter.h:
```diff
@@ -0,0 +1,67 @@
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+#ifndef STORAGE_ROCKSDB_INCLUDE_WAL_FILTER_H_
+#define STORAGE_ROCKSDB_INCLUDE_WAL_FILTER_H_
```
Now "#pragma once" is widely used even under include/rocksdb. Should follow it.

PraveenSinghRao replied:
Sure. I wasn't sure if it is supported everywhere. I can change to it.

siying commented on util/options.cc:
```diff
@@ -255,7 +256,11 @@ DBOptions::DBOptions()
-      wal_recovery_mode(WALRecoveryMode::kTolerateCorruptedTailRecords) {
+      wal_recovery_mode(WALRecoveryMode::kTolerateCorruptedTailRecords)
+#ifndef ROCKSDB_LITE
+      ,wal_filter(nullptr) 
```
Add a space after ","?

siying commented on db/db_impl.cc:
```diff
@@ -1153,6 +1154,28 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& log_numbers,
+        default:
+          assert(false); //unhandled case
+        }
+      }
```
Because of group commit, the write batch returned here might be multiple write batches grouped together. Is it OK? Do we want a finer granularity of control? For example, do we want to filter out record by record?

PraveenSinghRao replied:
IMO such multiple writes should be treated atomically, so current granularity is appropriate.

siying replied:
Just to double check, it's not about updating multiple keys in one write, but more concurrent writes by different threads. Is it good to you?

PraveenSinghRao replied:
If application itself commits a batch with LogData what would happen? Would there be multiple LogData in one batch here? If that is true, it might be good to break it down. How can we get each atomic batch from WriteBatch here?

siying replied:
From the source codes, it will be multiple LogData in one batch.

What you can do is to have the callback function to have an optional write batch returned. If it is not empty, we insert the returned write batch, instead of the original one. Would that work for you?

PraveenSinghRao replied:
I see. That should work - let me rework the code with this. Thanks.

siying commented on include/rocksdb/wal_filter.h:
```diff
@@ -0,0 +1,67 @@
+// of patent rights can be found in the PATENTS file in the same directory.
+// Copyright (c) 2013 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
```
Also perhaps we don't have to include the LevelDB headers in new files.

siying commented on include/rocksdb/wal_filter.h:
```diff
@@ -0,0 +1,67 @@
+#ifndef STORAGE_ROCKSDB_INCLUDE_WAL_FILTER_H_
+#define STORAGE_ROCKSDB_INCLUDE_WAL_FILTER_H_
+
+#ifndef ROCKSDB_LITE
```
Guard should be avoided in public header. Maybe just leave it in the header files.

PraveenSinghRao commented:
@dhruba @siying - addressed all the comments other than passing a vector of records to filter - I believe that would be a bit more intrusive change with the current code (since currently records are processed one by one) that can hopefully be done in the future if it would be significantly advantageous.

Please take a look when you get a chance and merge or provide any feedback. Thanks!

dhruba commented:
The public API looks fine to me, I am hoping that siying will get a chance to review it in detail

siying commented:
@dhruba I'm doing it now.

siying commented on db/db_impl.cc:
```diff
@@ -1155,6 +1156,73 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& log_numbers,
+              "Recovering log #%" PRIu64 " mode %d log filter %s returned " 
+              "more records (%d) than original (%d)", log_number,
+              db_options_.wal_recovery_mode, db_options_.wal_filter->Name(),
+              new_count, original_count);
```
This one is interesting. Now I think it is safer to fail.
If I remember correctly, we bump up sequence ID by one after each record: https://github.com/facebook/rocksdb/blob/v4.0/db/write_batch.cc#L464

Having more keys than we initially had can end up inserting keys whose sequence ID to be higher than later rows. It's may be something fixable but if you don't have a feature relying on it, I would simply fail.

PraveenSinghRao replied:
Sure, will fail.

siying commented on db/db_test.cc:
```diff
@@ -9889,6 +9890,341 @@ TEST_F(DBTest, PauseBackgroundWorkTest) {
+  // Create a test filter that would apply WalProcessingOption at the first
+  // record
+  size_t changeRecordsFromIndex = 1;
+  size_t numKeysToAddInNewBatch = 1;
```
Our naming convention is like change_records_from_index.

PraveenSinghRao replied:
Will fix.

siying commented on db/db_test.cc:
```diff
@@ -9889,6 +9890,341 @@ TEST_F(DBTest, PauseBackgroundWorkTest) {
+    // Index at which to start changing records
+    size_t m_changeRecordsFromIndex;
+    // Number of keys to add in the new batch
+    size_t m_numKeysToAddInNewBatch;
```
Tailing _.

siying commented on db/db_test.cc:
```diff
@@ -9889,6 +9890,341 @@ TEST_F(DBTest, PauseBackgroundWorkTest) {
+  class ChangeBatchHandler : public WriteBatch::Handler {
+  private:
+    // Batch to insert keys in
+    WriteBatch* newWriteBatch_;
```
new_write_batch_;

siying commented on db/db_test.cc:
```diff
@@ -9889,6 +9890,341 @@ TEST_F(DBTest, PauseBackgroundWorkTest) {
+  class TestWalFilter : public WalFilter {
+  private:
+    // Processing option that is requested to be applied at the given index
+    WalFilter::WalProcessingOption WalProcessingOption_;
```
wal_processing_option_

siying commented on db/db_test.cc:
```diff
@@ -9889,6 +9890,341 @@ TEST_F(DBTest, PauseBackgroundWorkTest) {
+    bool checkedAfterReopen = false;
+
+    while (true)
+    {
```
This needs to go to the line as while.

siying commented:
The patch looks good to me. There are some comments related to code convention. I can't list all of them, but need to fix that. Other than that, It's a great commit!

PraveenSinghRao commented:
I will fix the naming convention issues. For other formatting issues, is there an auto-formatter - when I see Travis build it seems to suggest formatting changes, it is possible to just apply them?

siying commented:
Don't worry too much about other formatting changes. Just do your best. I can fix them as a follow-up.

PraveenSinghRao commented:
Added the change to fail recovery if filter provides more records than original and corresponding unit-test. Fixed naming conventions.

Travis build failure doesn't seem to be related to my change.

Would you like me to squash the commits into one?

siying commented:
The test failure is odd. Let me try on my host and see how it works.

siying commented:
@PraveenSinghRao I tried to apply your pull request against master but there is conflicts. Can you resolve the conflicts?
